Add log

coko156 opened the pull request:
[+] New conf: mysqlnd_azure.logOutput, control write logs to nowhere/stdout/file
[m] fix a corruption.
[m] fix typo in doc and add some infomation.

santo-ms committed:
test

santo-ms committed:
test2

[+] Construct/Destruct inside MINIT/MSHUTDOWN MACROs
[+] php info
And some modifications.

santo-ms committed:
test3

[+] make logfilePath immutable
[+] change some function order in case of crash
[+] [SYSTM] level used by inner debug, not affected by loglevel

santo-ms committed:
test4

[+] Filename length restriction (255)
[+] Expose some logfile error to upper PHP module using php_error_docref()
[+] package.xml placeholder

santo-ms committed:
test5

[+] document
[M] README.md

coko156 committed:
Update mysqlnd_azure_log.md

[m] Fix some typos

santo-ms committed:
test6

[+] `[mysqlnd_azure] tag for error threw`

GuuBu committed:
Merge pull request #12 from microsoft/crash_alloc_prefix

fix crash due to wrong alloc pair

coko156 committed:
Update mysqlnd_azure_log.md

Typo fixed

GuuBu committed:
Merge branch 'master' into add_log

santo-ms committed:
test10

[+] New conf: mysqlnd_azure.logOutput, control write logs to nowhere/stdout/file
[m] fix a corruption.
[m] fix typo in doc and add some infomation.

dionysus1016 commented on utils.h:
```diff
@@ -0,0 +1,65 @@
+#define OPEN_LOGFILE(filename)                                                \
+  do {                                                                        \
+    if (filename != NULL) {                                                   \
+      logfile = fopen(filename, "a");                                         \
```
fopen will fail, which logfile will be null, take care of this when we write logs, I see you already checked this in close.

dionysus1016 commented on mysqlnd_azure.c:
```diff
@@ -674,6 +711,43 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
+      AZURE_LOG_SYS("Given logfile name too long, redirected to default: mysqlnd_azure_runtime.log");
+    }
+
+    if (logfile == NULL) {
```
move this to 727?

dionysus1016 commented on mysqlnd_azure.c:
```diff
@@ -674,6 +711,43 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
+
+/* {{{ mysqlnd_azure_release_resources, release resources when module destruct */
+int mysqlnd_azure_release_resources() {
+  if (MYSQLND_AZURE_G(logOutput) == 2 && MYSQLND_AZURE_G(logLevel) > 0 && logfile) {
```
we might could always call CLOSE_LOGFILE?

dionysus1016 replied:
never mind, just in case someone made more code changes in future and miss to change this code line.

coko156 replied:
Updated.
[+] Add some comments on code itself (resource_apply/ resource_release).
[m] open/close file handler when configured print logs to file
       logFile always opened if configured log to file, whatever the logLevel value is.

---
Former implementation may trigger some misleading behavior:
file opened but no close (if user change logLevel: 0 -> $some_number)
module exit without file close (if user change logLevel: $some_nuber -> 0)

santo-ms committed:
Add "[MYSQLND_AZURE]" header when print logs to stdout

santo-ms committed:
Change the behavior of log file handler open/close

santo-ms committed:
check if logfile available before write.

GuuBu commented:
/AzurePipelines run

santo-ms committed:
Write logs to stdout

When used as web server, message to stdout usually redirected to web
page. Redirect server logs to stderr is more suitable.

coko156 commented:
Tested with Apache:
![image](https://user-images.githubusercontent.com/4216136/85094299-0c679000-b221-11ea-9d5a-20566d26cb88.png)

coko156 commented:
Since [Apache Errror logs](https://httpd.apache.org/docs/2.4/logs.html) says:
"A very wide variety of different messages can appear in the error log. Most look similar to the example above. The error log will also contain debugging output from CGI scripts. Any information written to stderr by a CGI script will be copied directly to the error log."

So it will be fine to write log down to stderr.
1. guarantee console/apache/... can catch logs
2. no other application header added to the error messagem, which may be redundant.

santo-ms committed:
fflush(stderr) in case of log buffered after some redirection.

coko156 commented:
> I think it also need to flush stderr to get the log in time, how do you test? If request for multiple times, will the log be written in time?

resolved.

santo-ms committed:
Support multi-level logs to PHP application / FILE / stderr

[m] document
[m] logOutput level
[m] using macro instead of magic number
[-] remove irresponsible string compare function call

santo-ms committed:
Add tag in error message

santo-ms committed:
fix typo

GuuBu commented on doc/mysqlnd_azure_log.md:
```diff
@@ -0,0 +1,94 @@
+
+mysqlnd_azure => enableRedirect
+enableRedirect => preferred
+logfilePath => santotest.log
```
should be consistent with line 76

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -654,13 +689,15 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
 
                     ret = (*pconn)->m->connect(pconn, redirect_host, redirect_user, password, database, redirect_info->redirect_port, socket_or_pipe, mysql_flags);
                     if (ret == FAIL) {
+                        AZURE_LOG(ALOG_LEVEL_INFO, "Redirect cache miss");
```
This should bot be "Redirect cache miss" since cache already found, but "Use redirect cache failed to connect"

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -654,13 +689,15 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
                         mysqlnd_azure_remove_redirect_cache(username.s, hostname.s, port);
 
                         ret = (*pconn)->m->connect(pconn, hostname, username, password, database, port, socket_or_pipe, mysql_flags);
                     }
```
if ret = PASS, it will be a very useful log if we can add "Using cache to connect succeeded"

GuuBu commented on mysqlnd_azure.c:
```diff
@@ -674,6 +711,53 @@ MYSQLND_METHOD(mysqlnd_azure, connect)(MYSQLND * conn_handle,
+  if (MYSQLND_AZURE_G(logOutput) & ALOG_TYPE_FILE) {
+    char *logfilePath = NULL;
+    int logflag = 0;
+    if (ZSTR_LEN(MYSQLND_AZURE_G(logfilePath)) > 155) {
```
doc said 255 while it uses 155 here, it need to be consistent

GuuBu commented on utils.h:
```diff
@@ -0,0 +1,85 @@
+    }                                                                                        \
+} while (0)
+
+#define AZURE_LOG_SYS(format, ...)                                                           \
```
It will be nice if we can make the almost identical function into one
